File: src/main/java/org/logister/android/LogisterEvent.java
package org.logister.android;

import org.json.JSONObject;

import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;

/** Generic Logister event. Use LogisterClient convenience methods for common cases. */
public final class LogisterEvent {
    private final String eventType;
    private final String message;
    private final String level;
    private final String fingerprint;
    private final String occurredAt;
    private final Map<String, Object> context;
    private final Map<String, Object> attributes;

    private LogisterEvent(Builder builder) {
        this.eventType = builder.eventType;
        this.message = builder.message;
        this.level = builder.level;
        this.fingerprint = builder.fingerprint;
        this.occurredAt = builder.occurredAt;
        this.context = Collections.unmodifiableMap(new LinkedHashMap<>(builder.context));
        this.attributes = Collections.unmodifiableMap(new LinkedHashMap<>(builder.attributes));
    }

    public static Builder builder(String eventType, String message) {
        return new Builder(eventType, message);
    }

    public String getEventType() {
        return eventType;
    }

    public String getMessage() {
        return message;
    }

    public String getLevel() {
        return level;
    }

    public String getFingerprint() {
        return fingerprint;
    }

    public String getOccurredAt() {
        return occurredAt;
    }

    public Map<String, Object> getContext() {
        return context;
    }

    public Map<String, Object> getAttributes() {
        return attributes;
    }

    JSONObject toJson() throws Exception {
        JSONObject json = new JSONObject();
        json.put("event_type", eventType);
        putIfPresent(json, "message", message);
        putIfPresent(json, "level", level);
        putIfPresent(json, "fingerprint", fingerprint);
        putIfPresent(json, "occurred_at", occurredAt);

        for (Map.Entry<String, Object> entry : attributes.entrySet()) {
            putIfPresent(json, entry.getKey(), entry.getValue());
        }

        if (!context.isEmpty()) {
            json.put("context", new JSONObject(context));
        }

        return json;
    }

    static void putIfPresent(JSONObject json, String key, Object value) throws Exception {
        if (key != null && !key.isEmpty() && value != null) {
            json.put(key, value);
        }
    }

    public static final class Builder {
        private final String eventType;
        private final String message;
        private String level;
        private String fingerprint;
        private String occurredAt;
        private final Map<String, Object> context = new LinkedHashMap<>();
        private final Map<String, Object> attributes = new LinkedHashMap<>();

        private Builder(String eventType, String message) {
            if (eventType == null || eventType.trim().isEmpty()) {
                throw new IllegalArgumentException("eventType is required");
            }
            this.eventType = eventType;
            this.message = message;
        }

        public Builder level(String level) {
            this.level = level;
            return this;
        }

        public Builder fingerprint(String fingerprint) {
            this.fingerprint = fingerprint;
            return this;
        }

        public Builder occurredAt(String occurredAt) {
            this.occurredAt = occurredAt;
            return this;
        }

        public Builder context(String key, Object value) {
            if (key != null && !key.isEmpty() && value != null) {
                this.context.put(key, value);
            }
            return this;
        }

        public Builder context(Map<String, ?> context) {
            if (context != null) {
                for (Map.Entry<String, ?> entry : context.entrySet()) {
                    context(entry.getKey(), entry.getValue());
                }
            }
            return this;
        }

        public Builder attribute(String key, Object value) {
            if (key != null && !key.isEmpty() && value != null) {
                this.attributes.put(key, value);
            }
            return this;
        }

        public LogisterEvent build() {
            return new LogisterEvent(this);
        }
    }
}

File: src/main/kotlin/org/logister/android/LogisterEvent.kt
package org.logister.android

import java.util.Collections
import java.util.LinkedHashMap
import org.json.JSONObject

/** Generic Logister event. Use LogisterClient convenience methods for common cases. */
public class LogisterEvent private constructor(builder: Builder) {
    public val eventType: String = builder.eventType
    public val message: String? = builder.message
    public val level: String? = builder.level
    public val fingerprint: String? = builder.fingerprint
    public val occurredAt: String? = builder.occurredAt
    public val context: Map<String, Any> = Collections.unmodifiableMap(LinkedHashMap(builder.context))
    public val attributes: Map<String, Any> = Collections.unmodifiableMap(LinkedHashMap(builder.attributes))

    @Throws(Exception::class)
    internal fun toJson(): JSONObject {
        val json = JSONObject()
        json.put("event_type", eventType)
        putIfPresent(json, "message", message)
        putIfPresent(json, "level", level)
        putIfPresent(json, "fingerprint", fingerprint)
        putIfPresent(json, "occurred_at", occurredAt)

        for ((key, value) in attributes) {
            putIfPresent(json, key, value)
        }

        if (context.isNotEmpty()) {
            json.put("context", JSONObject(context))
        }

        return json
    }

    public class Builder internal constructor(
        internal val eventType: String,
        internal val message: String?
    ) {
        internal var level: String? = null
        internal var fingerprint: String? = null
        internal var occurredAt: String? = null
        internal val context: MutableMap<String, Any> = LinkedHashMap()
        internal val attributes: MutableMap<String, Any> = LinkedHashMap()

        init {
            require(eventType.isNotBlank()) { "eventType is required" }
        }

        public fun level(level: String?): Builder = apply {
            this.level = level
        }

        public fun fingerprint(fingerprint: String?): Builder = apply {
            this.fingerprint = fingerprint
        }

        public fun occurredAt(occurredAt: String?): Builder = apply {
            this.occurredAt = occurredAt
        }

        public fun context(key: String?, value: Any?): Builder = apply {
            if (!key.isNullOrEmpty() && value != null) {
                context[key] = value
            }
        }

        public fun context(context: Map<String, *>?): Builder = apply {
            context?.forEach { (key, value) ->
                context(key, value)
            }
        }

        public fun attribute(key: String?, value: Any?): Builder = apply {
            if (!key.isNullOrEmpty() && value != null) {
                attributes[key] = value
            }
        }

        public fun build(): LogisterEvent = LogisterEvent(this)
    }

    public companion object {
        @JvmStatic
        public fun builder(eventType: String, message: String?): Builder = Builder(eventType, message)

        @Throws(Exception::class)
        internal fun putIfPresent(json: JSONObject, key: String?, value: Any?) {
            if (!key.isNullOrEmpty() && value != null) {
                json.put(key, value)
            }
        }
    }
}

